server/subscribe.go

```diff
@@ -0,0 +1,472 @@
+			channelIds = append(channelIds, sub.ChannelId)
+		}
+	}
+
+	return channelIds, nil
+}
+
+func (p *Plugin) getSubscriptions() (*Subscriptions, error) {
+	data, err := p.API.KVGet(JIRA_SUBSCRIPTIONS_KEY)
+	if err != nil {
+		return nil, err
+	}
+	return SubscriptionsFromJson(data)
+}
+
+func (p *Plugin) getSubscriptionsForChannel(channelId string) ([]ChannelSubscription, error) {
+	subs, err := p.getSubscriptions()
+	if err != nil {
+		return nil, err
+	}
+
+	channelSubscriptions := []ChannelSubscription{}
+	for _, channelSubscriptionId := range subs.Channel.IdByChannelId[channelId] {
+		channelSubscriptions = append(channelSubscriptions, subs.Channel.ById[channelSubscriptionId])
+	}
+
+	return channelSubscriptions, nil
+}
+
+func (p *Plugin) removeChannelSubscription(subscriptionId string) error {
+	return p.atomicModify(JIRA_SUBSCRIPTIONS_KEY, func(initialBytes []byte) ([]byte, error) {
+		subs, err := SubscriptionsFromJson(initialBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		subscription, ok := subs.Channel.ById[subscriptionId]
+		if !ok {
+			return nil, errors.New("could not find subscription")
+		}
+
+		subs.Channel.remove(&subscription)
+
+		modifiedBytes, marshalErr := json.Marshal(&subs)
+		if marshalErr != nil {
+			return nil, marshalErr
+		}
+
+		return modifiedBytes, nil
+	})
+}
+
+func (p *Plugin) addChannelSubscription(newSubscription *ChannelSubscription) error {
+	return p.atomicModify(JIRA_SUBSCRIPTIONS_KEY, func(initialBytes []byte) ([]byte, error) {
+		subs, err := SubscriptionsFromJson(initialBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		newSubscription.Id = model.NewId()
+		subs.Channel.add(newSubscription)
+
+		modifiedBytes, marshalErr := json.Marshal(&subs)
+		if marshalErr != nil {
+			return nil, marshalErr
+		}
+
+		return modifiedBytes, nil
+	})
+}
+
+func (p *Plugin) editChannelSubscription(modifiedSubscription *ChannelSubscription) error {
+	return p.atomicModify(JIRA_SUBSCRIPTIONS_KEY, func(initialBytes []byte) ([]byte, error) {
+		subs, err := SubscriptionsFromJson(initialBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		oldSub, ok := subs.Channel.ById[modifiedSubscription.Id]
+		if !ok {
+			return nil, errors.New("Existing subscription does not exist.")
+		}
+		subs.Channel.remove(&oldSub)
+		subs.Channel.add(modifiedSubscription)
+
+		modifiedBytes, marshalErr := json.Marshal(&subs)
+		if marshalErr != nil {
+			return nil, marshalErr
+		}
+
+		return modifiedBytes, nil
+	})
+}
+
+func (p *Plugin) atomicModify(key string, modify func(initialValue []byte) ([]byte, error)) error {
+	readModify := func() ([]byte, []byte, error) {
+		initialBytes, appErr := p.API.KVGet(key)
+		if appErr != nil {
+			return nil, nil, errors.Wrap(appErr, "unable to read inital value")
+		}
+
+		modifiedBytes, err := modify(initialBytes)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "modification error")
+		}
+
+		return initialBytes, modifiedBytes, nil
+	}
+
+	success := false
+	for !success {
```

Comment:
1/5 should have a retry limit

Reply:
Want to make this a helper, will add fancy features like that then.